### PR TITLE
Implement improvements to density matrices.

### LINKED ISSUE #713: Basis-transforming 1- and 2-DMs
**Describe the feature you'd like**
I'd like to make sure that all 1- and 2-DM types can be basis-transformed.

**Describe what the current code offers in relation to what it lacks**
- [x] `SpinResolved1DM` uses explicit implementations for `transformed`: it should use `SpinResolvedBasisTransformable`;
- [x] `SpinResolved1DM` is missing conformance `SpinResolvedJacobiRotatable`;
- [x] `Simple1DM` is missing conformance to `JacobiRotatable`;
- [x] `Simple2DM` is missing conformance to `BasisTransformable`; (best tackled after #711 to use an easier API for the relevant contractions, as they're related to the formulas for `SimpleSQTwoElectronOperator`)
- [x] `Simple2DM` is missing conformance to `JacobiRotatable`;
- [x] `SpinResolved2DM` is missing conformance to `BasisTransformable`;
- [x] `SpinResolved2DM` is missing conformance to `JacobiRotatable`;
- [x] Nuke `OneDM` and check for uses of "OneDM";
- [x] Nuke `TwoDM` and check for uses of "TwoDM";
- [x] Fully remove `TransformationMatrix`;
- [x] Check for uses of `_Placeholder`.
- [x] Add a test for the transformed operators and density matrices.

Afterwards, a disabled test may be re-enabled:
~- [ ] re-enable `AtomicDecompositionParameters_test`;~

### PR DESCRIPTION
**Pull request by lelemmen**
This PR implements all features in #713. It closes #713.

**Don't forget:**
- [ ] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

**Comment by lelemmen** [comment]
I've added a small estimate to this PR itself too, because I had to implement 2-DMs for R/U/GHF and refactor common `DoublySpinResolved` transformation behavior.

@xdvriend I've asked for your review, once again.

File: gqcp/include/Operator/SecondQuantized/MixedUSQTwoElectronOperatorComponent.hpp [hunk 1]
```diff
@@ -127,15 +127,15 @@ class MixedUSQTwoElectronOperatorComponent:
         for (size_t i = 0; i < this->numberOfComponents(); i++) {
             switch (sigma) {
             case Spin::alpha: {
-                const auto temp = T_conjugate.template einsum<1>("UQ,TUVW->TQVW", parameters[i]);
-                const auto transformed = T_tensor.template einsum<1>("TP,TQVW->PQVW", temp);
+                const auto temp = T_tensor.template einsum<1>("UQ,TUVW->TQVW", parameters[i]);
+                const auto transformed = T_conjugate.template einsum<1>("TP,TQVW->PQVW", temp);
```
Comment (xdvriend) [inline gqcp/include/Operator/SecondQuantized/MixedUSQTwoElectronOperatorComponent.hpp:131]: Good catch on the fact that the indices which have to be transformed with the conjugate matrix were switched here!

File: gqcp/include/QCModel/HF/GHF.hpp [hunk 2]
```diff
@@ -240,10 +241,47 @@ class GHF {
...
+     *
+     *  @return The GHF 2-DM expressed in an orthonormal spinor basis.
+     */
+    static G2DM<Scalar> calculateOrthonormalBasis2DM(const size_t M, const size_t N) {
+
+        // Create the orbital space to determine the loops.
+        const auto orbital_space = GHF<Scalar>::orbitalSpace(M, N);
+
+
+        // Implement a KISS formula for the GHF 2-DM
```
Comment (xdvriend) [inline gqcp/include/QCModel/HF/GHF.hpp:265]:
```suggestion
        // Implement a KISS formula for the GHF 2-DM.
```

File: gqcp/include/QCModel/HF/RHF.hpp [hunk 3]
```diff
@@ -210,10 +212,54 @@ class RHF {
...
+        }
+
+        const size_t N_P = N / 2;  // The number of electron pairs.
+
+
+        // Create the orbital space to determine the loops.
+        const auto orbital_space = RHF<Scalar>::orbitalSpace(K, N_P);
+
+
+        // Implement a KISS formula for the GHF 2-DM.
```
Comment (xdvriend) [inline gqcp/include/QCModel/HF/RHF.hpp:243]:
```suggestion
        // Implement a KISS formula for the RHF 2-DM.
```

**Review by xdvriend — APPROVED** [review]
I noticed the GitHub action build fails due to something that's wrong with the `.transformed(U)` API. I didn't immediately spot the mistake during my review, So that still has to be addressed.

Other than that, most of my review comments are just nitpicking on punctuation and stuff like that. So once the mistake that causes the build to fail is fixed, this PR is good to go in my opinion.

File: gqcp/include/Operator/SecondQuantized/MixedUSQTwoElectronOperatorComponent.hpp (hunk 1 above)
Comment (lelemmen) [reply gqcp/include/Operator/SecondQuantized/MixedUSQTwoElectronOperatorComponent.hpp:131]: Yeah, I was re-checking the indices and I noticed something was off... :)

**Comment by lelemmen** [comment]
@xdvriend Thanks for your thorough review. A fresh pair of eyes is always welcome!

**Comment by xdvriend** [comment]
> @xdvriend Thanks for your thorough review. A fresh pair of eyes is always welcome!

No problem whatsoever! :smile:

**Pull request merged**